src/__test__/extra/write-after-end.test.ts

```diff
@@ -1,39 +1,96 @@
-import { createTestNode, delay, jsonTestEvents } from "@test-utils";
+/** @jest-environment ./src/__test__/utils/enableVersionCheck.ts */
+
+import {
+  createTestNode,
+  delay,
+  jsonTestEvents,
+  matchServerVersion,
+  optionalTest,
+} from "@test-utils";
 import { EventStoreDBClient, UnavailableError } from "@eventstore/db-client";
 
-describe("write after end", () => {
-  const node = createTestNode();
-  let client!: EventStoreDBClient;
+// These tests can take time.
+jest.setTimeout(120_000);
 
-  beforeAll(async () => {
+describe("write after end", () => {
+  test("Should not write after end", async () => {
+    const node = createTestNode();
     await node.up();
-    client = new EventStoreDBClient(
+
+    const client = new EventStoreDBClient(
       { endpoint: node.uri },
-      { rootCertificate: node.rootCertificate },
-      { username: "admin", password: "changeit" }
+      { rootCertificate: node.rootCertificate }
     );
-  });
 
-  afterAll(async () => {
-    await node.down();
-  });
-
-  test("Should not write after end", async () => {
     const STREAM_NAME = "json_stream_name";
-    await client.appendToStream(STREAM_NAME, jsonTestEvents());
-
-    client
-      .appendToStream(STREAM_NAME, jsonTestEvents(100_000))
-      .then((result) => {
-        expect(result).toBe("unreachable");
-      })
-      .catch((e) => {
-        expect(e).toBeInstanceOf(UnavailableError);
+    await client.appendToStream(STREAM_NAME, jsonTestEvents(), {
+      // credentials enforces classic append
+      credentials: { username: "admin", password: "changeit" },
+    });
+
+    const writeUntilError = () =>
+      new Promise((resolve) => {
+        const writeOnLoop = (): Promise<never> =>
+          client
+            .appendToStream(STREAM_NAME, jsonTestEvents(30_000), {
+              // credentials enforces classic append
+              credentials: { username: "admin", password: "changeit" },
+            })
+            .then(writeOnLoop);
+
+        writeOnLoop().catch((e) => {
+          resolve(e);
+        });
       });
 
-    node.killNode(node.endpoints[0]);
+    const errorPromise = writeUntilError();
+
+    await node.killNode(node.endpoints[0]);
+
+    const error = await errorPromise;
+    expect(error).toBeInstanceOf(UnavailableError);
 
     // wait for any unhandled rejections
     await delay(5_000);
+
+    await node.down();
   });
+
+  optionalTest(matchServerVersion`>=21.10`)(
+    "Should not write after end (batch append)",
+    async () => {
+      const node = createTestNode();
+      await node.up();
+
+      const client = new EventStoreDBClient(
+        { endpoint: node.uri },
+        { rootCertificate: node.rootCertificate }
```

Comment:
**CONCERN:** Shouldn't we set global credentials here?

Reply:
What's your reasoning?

Reply:
The other tests use credentials, so I think that we should be consistent. It'd be probably worth having a matrix for both calls with and without ACLs.

Reply:
> The other tests use credentials, so I think that we should be consistent.
Probably better not to be consistent in tests, as you're more likely to catch something when you do things differently?

> It'd be probably worth having a matrix for both calls with and without ACLs.

Credentials are specifically tested here:

https://github.com/EventStore/EventStore-Client-NodeJS/blob/master/src/__test__/connection/defaultCredentials.test.ts

This test file is ensuring that we don't have a regression on #208

Reply:
Inconsistency may be helpful, but better be consistent and proper matrix coverage 😉

Test that you linked checks only read methods it lacks checks for appends.

Reply:
> Test that you linked checks only read methods it lacks checks for appends.

That's more a general hole in the testing, rather than this PR. I'll open an issue for it.

Reply:
Issue is here #230